File: leetcode/0642-design-search-autocomplete-system/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class AutocompleteSystem {
public:
    struct TrieNode {
        unordered_map<char, TrieNode*> children;
        unordered_map<string, int> sentenceCounts;
    };

    TrieNode* root;
    unordered_map<string, int> freq;
    string currentPrefix;
    TrieNode* currentNode;
    bool fellOff;

    AutocompleteSystem(vector<string>& sentences, vector<int>& times) {
        root = new TrieNode();
        currentNode = root;
        fellOff = false;
        for (size_t i = 0; i < sentences.size(); i++) {
            freq[sentences[i]] = times[i];
            insert(sentences[i], times[i]);
        }
    }

    void insert(const string& sentence, int count) {
        TrieNode* node = root;
        for (char ch : sentence) {
            if (!node->children.count(ch)) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
            node->sentenceCounts[sentence] = count;
        }
    }

    vector<string> input(char c) {
        if (c == '#') {
            string sentence = currentPrefix;
            freq[sentence] = freq.count(sentence) ? freq[sentence] + 1 : 1;
            insert(sentence, freq[sentence]);
            currentPrefix.clear();
            currentNode = root;
            fellOff = false;
            return {};
        }

        currentPrefix += c;
        if (fellOff) {
            return {};
        }

        if (!currentNode->children.count(c)) {
            fellOff = true;
            return {};
        }

        currentNode = currentNode->children[c];
        vector<pair<string, int>> candidates(currentNode->sentenceCounts.begin(), currentNode->sentenceCounts.end());
        sort(candidates.begin(), candidates.end(), [](const pair<string, int>& a, const pair<string, int>& b) {
            if (a.second != b.second) {
                return a.second > b.second;
            }
            return a.first < b.first;
        });

        vector<string> result;
        for (size_t i = 0; i < candidates.size() && i < 3; i++) {
            result.push_back(candidates[i].first);
        }
        return result;
    }
};

File: leetcode/0642-design-search-autocomplete-system/Solution.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class AutocompleteSystem {
    class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        Map<String, Integer> sentenceCounts = new HashMap<>();
    }

    private final TrieNode root = new TrieNode();
    private final Map<String, Integer> freq = new HashMap<>();
    private StringBuilder currentPrefix = new StringBuilder();
    private TrieNode currentNode = root;
    private boolean fellOff = false;

    public AutocompleteSystem(String[] sentences, int[] times) {
        for (int i = 0; i < sentences.length; i++) {
            freq.put(sentences[i], times[i]);
            insert(sentences[i], times[i]);
        }
    }

    private void insert(String sentence, int count) {
        TrieNode node = root;
        for (char ch : sentence.toCharArray()) {
            node = node.children.computeIfAbsent(ch, k -> new TrieNode());
            node.sentenceCounts.put(sentence, count);
        }
    }

    public List<String> input(char c) {
        if (c == '#') {
            String sentence = currentPrefix.toString();
            freq.put(sentence, freq.getOrDefault(sentence, 0) + 1);
            insert(sentence, freq.get(sentence));
            currentPrefix = new StringBuilder();
            currentNode = root;
            fellOff = false;
            return new ArrayList<>();
        }

        currentPrefix.append(c);
        if (fellOff) {
            return new ArrayList<>();
        }

        if (!currentNode.children.containsKey(c)) {
            fellOff = true;
            return new ArrayList<>();
        }

        currentNode = currentNode.children.get(c);
        List<Map.Entry<String, Integer>> candidates = new ArrayList<>(currentNode.sentenceCounts.entrySet());
        candidates.sort((a, b) -> {
            if (!a.getValue().equals(b.getValue())) {
                return b.getValue() - a.getValue();
            }
            return a.getKey().compareTo(b.getKey());
        });

        List<String> result = new ArrayList<>();
        for (int i = 0; i < Math.min(3, candidates.size()); i++) {
            result.add(candidates.get(i).getKey());
        }
        return result;
    }
}
